### ISSUE TITLE
Docker killed due to memory leak

### ISSUE THREAD

**Issue by r4sas** [moby/moby#43262]
**Description**

Few days ago I updated my server system to Debian 10 and docker-ce was updated to 20.10.12~3. Today teammates reported that containers was killed few times. Further investigation shown that dockerd is leaking memory.

**Steps to reproduce the issue:**
Use Debian 10 and docker-ce 20.10.12~3 package.

**Describe the results you received:**
```
28215 root      20   0 4117436   2.3g  10028 S   2.0  58.8   0:14.64  `- /usr/bin/dockerd
```
dockerd uses from 70 MB of RAM and that value increases until oom error.

**Describe the results you expected:**
Normal usage of RAM

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client: Docker Engine - Community
 Version:           20.10.12
 API version:       1.41
 Go version:        go1.16.12
 Git commit:        e91ed57
 Built:             Mon Dec 13 11:45:37 2021
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.12
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.16.12
  Git commit:       459d0df
  Built:            Mon Dec 13 11:43:46 2021
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.12
  GitCommit:        7b11cfaabd73bb80907dd23182b9347b4245eb5d
 runc:
  Version:          1.0.2
  GitCommit:        v1.0.2-0-g52b36a2
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

**Output of `docker info`:**

```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Docker Buildx (Docker Inc., v0.7.1-docker)
  scan: Docker Scan (Docker Inc., v0.12.0)

Server:
 Containers: 5
  Running: 3
  Paused: 0
  Stopped: 2
 Images: 40
 Server Version: 20.10.12
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 7b11cfaabd73bb80907dd23182b9347b4245eb5d
 runc version: v1.0.2-0-g52b36a2
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: default
 Kernel Version: 4.19.0-18-rt-amd64
 Operating System: Debian GNU/Linux 10 (buster)
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 3.852GiB
 Name: <<redacted>>
 ID: <<redacted>>
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
```
Distributor ID: Debian
Description:    Debian GNU/Linux 10 (buster)
Release:        10
Codename:       buster

Linux <<redacted>> 4.19.0-18-rt-amd64 #1 SMP PREEMPT RT Debian 4.19.208-1 (2021-09-29) x86_64 GNU/Linux
```

Cloud VM with Intel(R) Xeon(R) CPU X5680  @ 3.33GHz

dockerd crash logs available here: https://paste.debian.net/hidden/e5738f18

**Comment by cmer81** [comment]
Hello,

i am also having the problem from the last update to 20.10.12

![image](https://user-images.githubusercontent.com/5068536/156170524-0893997f-4433-48b6-93b6-e79c4d361f94.png)

**Output of** `docker version`:

```
Client: Docker Engine - Community
 Version:           20.10.12
 API version:       1.41
 Go version:        go1.16.12
 Git commit:        e91ed57
 Built:             Mon Dec 13 11:45:33 2021
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.12
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.16.12
  Git commit:       459d0df
  Built:            Mon Dec 13 11:43:42 2021
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.12
  GitCommit:        7b11cfaabd73bb80907dd23182b9347b4245eb5d
 runc:
  Version:          1.0.2
  GitCommit:        v1.0.2-0-g52b36a2
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
  ```

**Output of** `docker info:`

```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Docker Buildx (Docker Inc., v0.7.1-docker)
  scan: Docker Scan (Docker Inc., v0.12.0)

Server:
 Containers: 15
  Running: 15
  Paused: 0
  Stopped: 0
 Images: 9
 Server Version: 20.10.12
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runtime.v1.linux runc io.containerd.runc.v2
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 7b11cfaabd73bb80907dd23182b9347b4245eb5d
 runc version: v1.0.2-0-g52b36a2
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: default
 Kernel Version: 5.4.0-100-generic
 Operating System: Ubuntu 20.04.4 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 8
 Total Memory: 15.51GiB
 Name: game
 ID: 76BT:QKP4:CTVU:UYD5:7BTX:X4X4:QV6N:KEY7:4DJY:MCCX:PW5V:J5UU
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

```
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 20.04.4 LTS
Release:        20.04
Codename:       focal
```

dedicated server at OVH 8x

```
Intel(R) Core(TM) i7-4790K CPU @ 4.00GHz
16Go memory 
5.4.0-100-generic
```

**Comment by olljanat** [comment]
@r4sas @cmer81 from which version you did upgrade to latest one? Just wondering that where to start looking for that what has changed.

**Comment by r4sas** [comment]
@olljanat for me latest version before upgrading system from Debian Stretch was `docker-ce_19.03.9~3-0`.

**Comment by olljanat** [comment]
Is this btw duplicate with #43043 which was fixed on just released [v20.10.13](https://github.com/moby/moby/releases/tag/v20.10.13) version?

**Comment by r4sas** [comment]
I'll check if anything changed.

**Comment by r4sas** [comment]
Everything looks better now, but `docker stats` is broken...

**Comment by r4sas** [comment]
@olljanat Everything looks better now, but `docker stats` is broken...

**Comment by maxime1907** [comment]
Its fixed with Docker version 20.10.13, build a224086, but docker stats is broken
![image](https://user-images.githubusercontent.com/19607336/158080968-30aa9c51-271f-47ba-a541-2d531ed5d4e2.png)

**Comment by olljanat** [comment]
Please, open another issue about that stats bug (if not yet reported by someone) as Moby maintainers not necessarily notice it from here.
